### PR TITLE
Fix wrong recognition for "10/1 - 11/7"

### PR DESCRIPTION
**Pull request by guom08**
#819 
The main cases handle in this PR includes:
1. Wrong splitting of DateRange entity, like "10/1-11/5", the original design would extract "10/1-11" as a date, which is wrong
2. Shared date context missing, like "Jan - March 2015", where "2015" should be the context of year.

Besides, there are some known issues:
1. "Dec/2018-Dec/2019", this still fails as we don't support recognizing "Dec/2018" for now #944
2. "10/1-10/2/2017", this still fails as after the Regex match "10/1-10", it will skip the "10/2/2017" as there is some overlapping #945
3. The Javascript and Python support is not in this PR, specs for other languages are also not in this PR as time is limited #946

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 1]
```diff
@@ -1919,5 +1951,52 @@ public List<DateTimeParseResult> FilterResults(string query, List<DateTimeParseR
...
+            var relativeMatchForEndDate = this.config.RelativeRegex.Match(endDateStr);
+            isDateRelative = relativeMatchForStartDate.Success || relativeMatchForEndDate.Success;
+
+            var dateContext = new DateContext();
+
+            if (!isEndDatePureYear && !isDateRelative)
+            {
+                foreach (Match match in config.YearRegex.Matches(text))
+                {
+                    var year = ((BaseDateExtractor)this.config.DateExtractor).GetYearFromText(match);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: This type of cast shouldn't be needed. If the method exists in the base class for all datetime extractors, why nod just add it to the common interface? Or make is static?

File: Specs/DateTime/English/DateTimeModel.json [hunk 2]
```diff
@@ -7468,5 +7468,461 @@
         }
       }
     ]
+  },
+  {
+    "Input": "My vacation is from 10-1-2018-10-7-2018",
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModel.json]: Can you please add all the corresponding specs to EnglishOther? We need to make sure there is no break between DMY and MDY date formats.

File: .NET/Microsoft.Recognizers.Text.DateTime/Utilities/DateContext.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            var startDate = SetDateWithContext(originalDateRange.Item1);
+            var endDate = SetDateWithContext(originalDateRange.Item2);
+
+            return new Tuple<DateObject, DateObject>(startDate, endDate);
+        }
+
+        private string SetTimexWithContext(string originalTimex)
+        {
+            return originalTimex.Replace(Constants.TimexFuzzyYear, Year.ToString("D4"));
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Utilities/DateContext.cs]: All timex manipulation code should go to the timex generation code. Let's try to not leave it scattered around.

File: .NET/Microsoft.Recognizers.Text.DateTime/Spanish/Extractors/SpanishDateExtractorConfiguration.cs [hunk 4]
```diff
@@ -67,11 +67,11 @@ public class SpanishDateExtractorConfiguration : BaseOptionsConfiguration, IDate
 
         public static readonly Regex RelativeMonthRegex = new Regex(DateTimeDefinitions.RelativeMonthRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
 
-        public static readonly Regex PrefixArticleRegex =
-            new Regex(
-                DateTimeDefinitions.PrefixArticleRegex,
+        public static readonly Regex PrefixArticleRegex = new Regex(DateTimeDefinitions.PrefixArticleRegex,
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Spanish/Extractors/SpanishDateExtractorConfiguration.cs]: Nit: standard

File: .NET/Microsoft.Recognizers.Text.DateTime/Spanish/Extractors/SpanishDateExtractorConfiguration.cs [hunk 5]
```diff
@@ -67,11 +67,11 @@ public class SpanishDateExtractorConfiguration : BaseOptionsConfiguration, IDate
 
         public static readonly Regex RelativeMonthRegex = new Regex(DateTimeDefinitions.RelativeMonthRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
 
-        public static readonly Regex PrefixArticleRegex =
-            new Regex(
-                DateTimeDefinitions.PrefixArticleRegex,
+        public static readonly Regex PrefixArticleRegex = new Regex(DateTimeDefinitions.PrefixArticleRegex,
                 RegexOptions.IgnoreCase | RegexOptions.Singleline);
 
+        public static readonly Regex RangeConnectorSymbolRegex = new Regex(Definitions.BaseDateTime.RangeConnectorSymbolRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Spanish/Extractors/SpanishDateExtractorConfiguration.cs]: Nit: standard

File: .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/PortugueseDatePeriodParserConfiguration.cs [hunk 6]
```diff
@@ -79,10 +79,12 @@ public class PortugueseDatePeriodParserConfiguration : BaseOptionsConfiguration,
         public static readonly Regex NextPrefixRegex = new Regex(DateTimeDefinitions.NextPrefixRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
         public static readonly Regex PastPrefixRegex = new Regex(DateTimeDefinitions.PastPrefixRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
         public static readonly Regex ThisPrefixRegex = new Regex(DateTimeDefinitions.ThisPrefixRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/PortugueseDatePeriodParserConfiguration.cs]: Nit: standard. All of them. :)

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 7]
```diff
@@ -1079,13 +1106,18 @@ private DateTimeResolutionResult MergeTwoTimePoints(string text, DateObject refe
...
+            var dateContext = GetYearContext(er[0].Text, er[1].Text, text);
+
             var pr1 = this.config.DateParser.Parse(er[0], referenceDate);
             var pr2 = this.config.DateParser.Parse(er[1], referenceDate);
             if (pr1.Value == null || pr2.Value == null)
             {
                 return ret;
             }
 
+            pr1 = dateContext.ProcessDateEntityParsingResult(pr1);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: "Process" for what?

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 8]
```diff
@@ -1079,13 +1106,18 @@ private DateTimeResolutionResult MergeTwoTimePoints(string text, DateObject refe
                 er[1].Text = weekPrefix + " " + er[1].Text;
             }
 
+            var dateContext = GetYearContext(er[0].Text, er[1].Text, text);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Is there a way to know if we really need to extract context earlier and skip this call if not?

File: .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDateExtractorConfiguration.cs [hunk 9]
```diff
@@ -150,6 +150,8 @@ public class ItalianDateExtractorConfiguration : BaseOptionsConfiguration, IDate
         public static readonly Regex RangeUnitRegex =
             new Regex(DateTimeDefinitions.RangeUnitRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
 
+        public static readonly Regex RangeConnectorSymbolRegex = new Regex(Definitions.BaseDateTime.RangeConnectorSymbolRegex, RegexOptions.IgnoreCase | RegexOptions.Singleline);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Italian/Extractors/ItalianDateExtractorConfiguration.cs]: Nit: standard

File: .NET/Microsoft.Recognizers.Text.DateTime/German/Parsers/GermanDatePeriodParserConfiguration.cs [hunk 10]
```diff
@@ -87,10 +87,15 @@ public class GermanDatePeriodParserConfiguration : BaseOptionsConfiguration,IDat
             new Regex(
                 DateTimeDefinitions.ThisPrefixRegex,
                 RegexOptions.IgnoreCase | RegexOptions.Singleline);
+        public static readonly Regex RelativeRegex =
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/German/Parsers/GermanDatePeriodParserConfiguration.cs]: Nit: standard

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs [hunk 11]
```diff
@@ -104,14 +105,96 @@ private List<Token> BasicRegexMatch(string text)
...
+                if (startSymbolLength > 0)
+                {
+                    text = text.Substring(startSymbolLength);
+                }
+            }
+
+            return text.Trim();
+        }
+
+        private bool StartsWithBasicDate(string text)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs]: Both TrimStartRangeConnectorSymbols and StartsWithBasicDate seem a little convoluted and use many regexes. 
For now I suggest adding a @TODO in the code to revisit later to simplify.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 7 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: This is to update the ParsingResult to use the date context (currently only "year" information is applied)

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 8 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Sorry, I didn't quite get the point.

**Pull request merged**